### PR TITLE
Add an Mmap wrapper to rustc_data_structures

### PR DESCRIPTION
**Pull request by bjorn3**
This wrapper implements StableAddress and falls back to directly reading the file on wasm32.

Taken from #83640, which I will close due to the perf regression.

### CODE REVIEW

**Comment by bjorn3** [comment]
It seemed to be a tiny (<0.1%) regression on the other PR. I can to benchmark after this commit.

**Comment by bjorn3** [comment]
This seems to be a slight regression on a few benchmarks. Trying with a few methods inlined.

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
That seems to be slightly better. Now trying with 8fcd03e7358ccb8f568ab3b34f3f31fac4b6ede9 cherry-picked.

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
Nope, that is a regression. Maybe due to the double `metadata` call by both this code and `memmap2`? I reverted it again. Ready for review.

File: compiler/rustc_data_structures/src/memmap.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+impl Deref for Mmap {
+    type Target = [u8];
+
+    #[inline]
+    fn deref(&self) -> &[u8] {
+        &*self.0
+    }
+}
+
+unsafe impl StableAddress for Mmap {}
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/memmap.rs:43]: r=me with a comment about the safety in both memmap and wasm cases.

**Comment by bjorn3** [comment]
Added the safety comment.

@bors r=cjgillot

**Pull request merged**